openghg/combine.py

```diff
@@ -0,0 +1,191 @@
+    return combine_data_objects(data_objects, preprocess=preprocess)
+
+
+def _data_array_from_value(
+    value: Any, coords: Union[xr.Coordinates, Mapping[Hashable, xr.DataArray]], name: Optional[str] = None
+) -> xr.DataArray:
+    """Create xr.DataArray with single value and given coords and name.
+
+    Args:
+        value: value to broadcast over coordinates (i.e. the constant value the DataArray should hold)
+        coords: coordinates from an xr.Dataset or xr.DataArray
+        name: name for output xr.DataArray
+
+    Returns:
+        xr.DataArray with single value for all coordinate values
+    """
+    shape = tuple(len(coord) for coord in coords.values())
+    data = np.broadcast_to(value, shape)
+    return xr.DataArray(data, coords=coords, name=name)
+
+
+def add_variable_from_metadata(
+    data_object: T,
+    metadata_key: str,
+    dims: Union[str, list[str]] = "time",
+    formatter: Callable = str,
+    drop_metadata_key: bool = True,
+    name: Optional[str] = None,
+) -> T:
+    """Add a data variable `new_dim `to the data of T using the value corresponding
+    to the key `new_dim` in the metadata of T.
+
+    Args:
+        data_object: object to modify (the metadata is copied; the data is a view?)
+        metadata_key: key of value that will be retrieved from data_object.metadata.
+        dims: dimension or list of dimensions the new variable should have
+        formatter: function to apply to retrieved metadata value. Only applied if value is not None.
+        drop_metadata_key: if True, del `metadata_key` from the metadata of the returned object.
+        name: name for the new variable; if `None` then `metadata_key` will be used.
+
+    Returns:
+        object of same type as `data_object` whose data has an data variable with given coordinates and constant
+        value, and whose metadata is the same, possibly with the value of the new dimension removed.
+    """
+    value = data_object.metadata.get(metadata_key, None)
+    if value and formatter is not None:
+        value = formatter(value)
+
+    if isinstance(dims, str):
+        dims = [dims]
+
+    coords = {k: v for k, v in data_object.data.coords.items() if k in dims}
+    if name is None:
+        name = metadata_key
+    new_da = _data_array_from_value(value, coords, name)
+
+    new_data = xr.merge([data_object.data, new_da])
+
+    new_metadata = data_object.metadata.copy()
+
+    if drop_metadata_key and metadata_key in new_metadata:
+        del new_metadata[metadata_key]
+
+    # make a new data object of same type as input
+    cls = type(data_object)
+    result = cls(metadata=new_metadata, data=new_data)
+
+    return result
+
+
+def combine_and_elevate_inlet(data_objects: list[T]) -> T:
+    """Combine multiple data objects into a single data object, elevating
+    the inlet value in their metadata to a data variable with a time dimension.
+    """
+
+    def inlet_formatter(inlet: str) -> float:
```

Comment:
I added a function `extract_float` to the strings utils. If no float is found, it raises a value error, so all the preprocess function does is catch the value error after calling that function. I didn't want to return `np.nan` by default in the general case, and since `float("nan")` is a valid float value, returning `float("nan")` wasn't a good way to indicate it failed, so I raised the error to leave it up to the caller. I could probably change this, but it will extract `float("nan")` from the string "banana"... I guess I could say "find 'nan' only if it isn't preceded by another letter", but I'm inclined to leave it

Reply:
Actually I've updated this now to check that 'nan', 'inf', and 'infinity' are not preceded or followed by a letter.